src/contentScript.js

```diff
@@ -33,51 +33,100 @@ async function freezeThisPage(options) {
  * Top-level dispatcher for commands sent from the devpanel or corpus collector
  * to this content script
  */
-async function dispatch(request) {
+function dispatch(request) {
     switch (request.type) {
+        case 'init':
+            injectSimmer();
+            break;
+
         case 'label':
-            // TODO: Show a message or something if no element is inspected.
-            elementAtPath(request.inspectedElement, document).setAttribute('data-fathom', request.label);
+            setLabel(request);
             break;
+
         case 'freeze':
-            // Freeze a page at the bidding of the corpus collector or devtools
-            // save button. Devtools panel calls this indirectly, by way of the
-            // background script, so it can download the result when done.
-            // Corpus collector calls directly.
-            if (request.inspectedElement !== undefined) {
-                hideHighlight();
-            }
-            const html = await freezeThisPage(request.options);
-            if (request.inspectedElement !== undefined) {
-                showHighlight(elementAtPath(request.inspectedElement, document));
-            }
-            return Promise.resolve(html);
+            return freezePage(request);
+
         case 'showHighlight':
-            if (request.inspectedElement !== undefined) {
-                showHighlight(elementAtPath(request.inspectedElement, document));
-            }
+            showHighlight(request.selector);
             break;
+
         case 'hideHighlight':
             hideHighlight();
             break;
     }
-    return Promise.resolve({});
 }
 browser.runtime.onMessage.addListener(dispatch);
 
+// Inject the Simmer library into the current page, so we can use it to
```

Comment:
Can we avoid injecting it? If we inject it into the DOM tree proper, our script tag will show up in all our freezes. Of course, it won't be functional or throw any errors, but I try to keep the page as stock as possible, because who knows what rulesets might draw on for signal?

I was going to suggest including simmer as another content script (in manifest.json), but I see you need it from the eval in the devpanel opener. Do you have any ideas? Worst case, I guess we could remove it before freezing and re-add it, but ick.

Reply:
yeah; i tried many approaches, this was the last on my list.  will look at removing during freezing.